### ISSUE TITLE
Better C interface

### ISSUE THREAD

**Issue by be5invis** [googlei18n/compreffor#20]
The current C interface lacks some import parts, we still need Python to deal with it.

**Comment by behdad** [comment]
If you mean the parts decoding SFNT and CFF table, then that's a lot of code; there's not much point reimplementing that in this module.  Maybe FreeType allows loading those?

**Comment by be5invis** [comment]
@behdad I just care about the subroutinizer algorithm. I have already a decoder and an encoder, all parts (like - the charstring INDEX) are avaliable. There is even an IL for charstrings.

**Comment by behdad** [comment]
I haven't looked at the code in a long time.  Can you summarize what changes you like to see?

**Comment by be5invis** [comment]
@behdad A simple interface, pretty like the existing
```c++
extern "C" uint32_t* compreff(unsigned char* dataStream, int numRounds, unsigned& outputLength);
```
However, instead of returning a “result” which is needed to decode, provide a more “direct” output like this:
```c++
typedef struct {
    uint32_t length;
    uint8_t  data;
} CFFIndex;

typedef struct {
    CFFIndex *charStrings;
    CFFIndex *globalSubroutines;
    uint32_t NumOfLocalSubroutines;
    CFFIndex **localSubroutines;
} ComprefforResult;

extern "C" ComprefforResult *compreff(const CFFIndex *inputCharStrings, uint32_t rounds);
```

**Comment by behdad** [comment]
I see.  Yes, I think we can do something like that.

**Comment by be5invis** [comment]
@behdad I’d like your C interface’s input and output plain enough. That is, can be used directly when building a CFF font.
And prehaps, you can drop the `fdSelect` part, and use one unified local subroutine table for ALL subfonts in a CID FDArray.

**Comment by behdad** [comment]
> And prehaps, you can drop the fdSelect part, and use one unified local subroutine table for ALL subfonts in a CID FDArray. In my case, all input charstrings do not contain any subroutine call.

Sure that's possible but comes at a theoretical cost.  Not important for most fonts though, I agree.

**Comment by be5invis** [comment]
@behdad I think that trade-off is acceptable.
